### PR TITLE
enhancement(remap): VRL web RFC

### LINKED ISSUE #6600: RFC for running Vector Remap Language in web environments
As we've learned throughout this project, [WebAssembly][wasm] creates a wide variety of intriguing possibilities for running Rust code in atypical environments (in the browser, on the edge, etc.). On Discord and Slack the Vector team has had some lively conversations on directions we could go with using one crucial part of Vector, [Vector Remap Language][vrl], in web environments. Numerous possibilities have been put on the table, none of which would involve a great deal of work given the strength of the Rust + Wasm ecosystem and the tools available. I propose an RFC for forming consensus around which direction(s) we should take this.

[vrl]: https://vrl.dev
[wasm]: https://webassembly.org

### PR DESCRIPTION
**Pull request by lucperkins**
This RFC provides some proposals for how we can eventually respond with an emphatic "YES" when people ask us "is VRL web yet?"

Closes #6600

### CODE REVIEW

File: rfcs/2021-03-02-6600-vector-remap-language-web.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+* The ability to share URLs for specific VRL events and/or programs. This would likely require using
+  a backend [web server] although it *may* be feasible to use URL hashes to achieve this.
+
+#### Prior art
+
+In PR [6590] I created a proof of concept for a VRL web app that uses the [Trunk] build tool and the
+[Yew] front-end framework for Rust, which is inspired by [React.js][react] and enables you to write
+web apps in pure Rust and without touching JavaScript. This POC does "work" but it creates an
+unacceptably large Wasm binary (~15 MB), largely due to it containing both VRL (which is bulky in
+itself) and *all* of the app's front-end logic. You can see a demo of that proof of concept
```
Comment (blt) [inline rfcs/2021-03-02-6600-vector-remap-language-web.md:40]: I got curious about this and ran [cargo bloat](https://github.com/RazrFalcon/cargo-bloat) on `lib/vrl/cli` at 386e1ea64f50f17d93f0f6ef4cd519903495c76c: 

```
➜  cli git:(master) ✗ cargo bloat --release --crates
    Finished release [optimized] target(s) in 0.13s
    Analyzing target/release/vrl

 File  .text     Size Crate
 7.9%  23.0% 788.7KiB std
 4.4%  12.9% 440.3KiB vrl_stdlib
 3.7%  10.8% 370.5KiB vrl_parser
 3.1%   9.0% 306.8KiB vrl_compiler
 2.5%   7.2% 246.4KiB clap
 1.9%   5.6% 192.3KiB regex
 1.8%   5.2% 176.6KiB regex_syntax
 1.3%   3.9% 133.7KiB [Unknown]
 1.1%   3.2% 110.4KiB rustyline
 0.9%   2.5%  85.7KiB aho_corasick
 0.6%   1.7%  59.5KiB chrono
 0.5%   1.4%  46.7KiB url
 0.4%   1.2%  39.8KiB vrl_cli
 0.4%   1.1%  37.2KiB term
 0.4%   1.0%  35.6KiB serde_json
 0.3%   1.0%  34.2KiB onig_sys
 0.3%   0.8%  27.8KiB hashbrown
 0.3%   0.8%  26.4KiB sha2
 0.2%   0.7%  24.3KiB codespan_reporting
 0.2%   0.7%  22.9KiB vrl_diagnostic
 1.9%   5.6% 192.2KiB And 41 more crates. Use -n N to show more.
34.2% 100.0%   3.3MiB .text section size, the file size is 9.8MiB

Note: numbers above are a result of guesswork. They are not 100% correct and never will be.
```

If we optimize for size in the top-level `Cargo.toml` we get:

```
    Finished release [optimized] target(s) in 1m 47s
    Analyzing target/release/vrl

 File  .text     Size Crate
 4.5%  20.2% 418.7KiB std
 3.1%  14.0% 290.2KiB vrl_stdlib
 2.3%  10.6% 220.2KiB vrl_parser
 2.1%   9.3% 192.2KiB vrl_compiler
 1.7%   7.5% 155.3KiB clap
 1.6%   7.5% 154.6KiB regex
 0.9%   4.2%  86.8KiB regex_syntax
 0.9%   4.1%  84.5KiB [Unknown]
 0.8%   3.7%  75.8KiB vrl_cli
 0.5%   2.2%  45.5KiB aho_corasick
 0.4%   1.8%  37.0KiB rustyline
 0.3%   1.5%  31.5KiB term
 0.3%   1.4%  28.5KiB url
 0.3%   1.3%  27.4KiB chrono
 0.3%   1.2%  25.1KiB vrl
 0.3%   1.2%  24.1KiB onig_sys
 0.2%   0.9%  18.0KiB sha2
 0.2%   0.8%  17.5KiB serde_json
 0.2%   0.8%  17.0KiB chrono_tz
 0.1%   0.6%  12.8KiB codespan_reporting
 1.1%   5.0% 103.2KiB And 54 more crates. Use -n N to show more.
22.1% 100.0%   2.0MiB .text section size, the file size is 9.2MiB

Note: numbers above are a result of guesswork. They are not 100% correct and never will be.
```

That's a little better. The `structopt` crate brings in clap which [can be big](https://github.com/clap-rs/clap/issues/1365). Anyhow, seems like if we _did_ want to slim this down we might consider if we can trim from the above. Function sizes:

```
➜  cli git:(master) ✗ cargo bloat -n 30 --split-std --release 
    Finished release [optimized] target(s) in 0.14s
    Analyzing target/release/vrl

 File  .text     Size        Crate Name
 1.5%   6.7% 139.7KiB   vrl_parser vrl_parser::parser::__parse__Program::__reduce
 0.4%   1.7%  34.5KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.3%   1.3%  26.7KiB    addr2line addr2line::Context<R>::from_dwarf
 0.3%   1.2%  24.1KiB        regex regex::exec::ExecBuilder::build
 0.2%   1.0%  19.9KiB    [Unknown] match_at
 0.2%   0.9%  18.8KiB   vrl_stdlib vrl_stdlib::parse_aws_alb_log::parse_log
 0.2%   0.9%  18.3KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.2%   0.9%  18.1KiB         clap clap::app::parser::Parser::get_matches_with
 0.2%   0.8%  16.6KiB    chrono_tz <chrono_tz::timezones::Tz as chrono_tz::timezone_impl::TimeSpans>::timespans
 0.2%   0.8%  16.0KiB vrl_compiler vrl_compiler::type_def::KindInfo::merge
 0.2%   0.7%  15.3KiB         clap clap::app::validator::Validator::validate
 0.2%   0.7%  14.9KiB vrl_compiler <vrl_compiler::expression::op::Op as vrl_compiler::expression::Expression>::resolve
 0.2%   0.7%  14.6KiB      vrl_cli vrl_cli::cmd::cmd
 0.1%   0.6%  13.3KiB    addr2line addr2line::ResUnit<R>::parse_lines
 0.1%   0.6%  12.6KiB      vrl_cli rustyline::readline_edit
 0.1%   0.6%  11.9KiB          std std::sys::unix::process::process_inner::<impl std::sys::unix::process::process_common::Command>::spawn
 0.1%   0.6%  11.9KiB          url url::host::Host::parse
 0.1%   0.6%  11.8KiB        regex <regex::re_trait::Matches<R> as core::iter::traits::iterator::Iterator>::next
 0.1%   0.5%  11.4KiB      vrl_cli <&T as core::fmt::Display>::fmt
 0.1%   0.5%  11.4KiB aho_corasick aho_corasick::packed::api::Builder::build
 0.1%   0.5%   9.9KiB         term term::terminfo::TermInfo::from_name
 0.1%   0.5%   9.7KiB         clap clap::app::help::Help::write_arg
 0.1%   0.5%   9.5KiB         sha2 sha2::sha512::compress512
 0.1%   0.5%   9.5KiB   vrl_stdlib <vrl_stdlib::parse_syslog::ParseSyslogFn as vrl_compiler::expression::Expression>::resolve
 0.1%   0.4%   9.1KiB   vrl_stdlib <vrl_stdlib::parse_aws_cloudwatch_log_subscription_message::ParseAwsCloudWatchLogSubscriptionMessageFn as vrl_compiler::expression::Expression>::resolve
 0.1%   0.4%   9.1KiB      vrl_cli rustyline::command::execute
 0.1%   0.4%   8.8KiB    [Unknown] parse_exp
 0.1%   0.4%   8.6KiB   vrl_stdlib vrl_stdlib::parse_aws_vpc_flow_log::parse_log
 0.1%   0.4%   8.4KiB  miniz_oxide miniz_oxide::inflate::core::decompress
 0.1%   0.4%   8.4KiB         core core::fmt::float::float_to_decimal_common_shortest
16.1%  73.1%   1.5MiB              And 8696 smaller methods. Use -n N to show more.
22.1% 100.0%   2.0MiB              .text section size, the file size is 9.2MiB
```

File: rfcs/2021-03-02-6600-vector-remap-language-web.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  the server stateful and using some kind of persistent data store or devising a system of inferring
+  events/programs via hashes passed in via URL. It's likely that those URLs would grow quite
+  unwieldy, however, so we should experiment before taking this route.
+
+### Published module
+
+Yet another possibility worth considering would be publishing Wasm modules to hubs like [npm] and
+[Wasmer] that others can use in their own environments. Those modules could be released alongside
+versions of VRL. Tools from the Rust + Wasm ecosystem, like [wasm-pack], make this quite simple in
+principle. With published modules, users could run VRL
```
Comment (blt) [inline rfcs/2021-03-02-6600-vector-remap-language-web.md:109]: Did the sentence end early or is a period missing?

File: rfcs/2021-03-02-6600-vector-remap-language-web.md (hunk 1 above)
Comment (lucperkins) [reply rfcs/2021-03-02-6600-vector-remap-language-web.md:40]: I should note here that a Wasm binary wouldn't need the CLI at all, and actually only requires `core` and `stdlib`. Here's what I get running this in `lib/vrl/wasm` (in the PR branch):

```
 File  .text     Size Crate
55.2% 100.1% 205.6KiB std
 0.2%   0.4%     752B wasm_bindgen
 0.0%   0.0%      64B [Unknown]
55.2% 100.0% 205.5KiB .text section size, the file size is 372.2KiB
```

File: rfcs/2021-03-02-6600-vector-remap-language-web.md (hunk 2 above)
Comment (lucperkins) [reply rfcs/2021-03-02-6600-vector-remap-language-web.md]: Excellent catch 👍🏼

File: rfcs/2021-03-02-6600-vector-remap-language-web.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+I propose pursuing #1 and #3 in the near term, while leaving #2 as an option for the future.
+
+### Interactive web app
+
+An interactive VRL web app would be a browser-based app that provides an aesthetically pleasing and
+UX-friendly experience for learning and experimenting with VRL. In its most basic form, it would
+enable you to enter a VRL event and a VRL program into a text area and see what happens. But there
+are other capabilities that we should consider:
+
```
Comment (sghall) [inline rfcs/2021-03-02-6600-vector-remap-language-web.md:26]: Just wondering if this be successful without the ability to select different versions of VRL?  Would this just be the latest released version?  Other related examples like the [rust playground](https://play.rust-lang.org) and [TS Playground](https://www.typescriptlang.org/play) allow you to select different versions including the nightly version. Of course that has implications for how we build this.

File: rfcs/2021-03-02-6600-vector-remap-language-web.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+* A step-by-step tutorial, like the one under way in PR [6184] that provides real exercises with
+  right and wrong answers
+* A built-in "library" of example events (JSON, Syslog, AWS formats, etc.)
```
Comment (sghall) [inline rfcs/2021-03-02-6600-vector-remap-language-web.md:29]: Yes, I see more "examples" that are selectable via drop down in the other playgrounds out there.  This is a lot simpler to implement and less friction for the user vs a step-by-step tutorial.  You select an example and loads up the text boxes with the input event and a sample script.  You can put inline comments in the example to explain what is being demonstrated.

File: rfcs/2021-03-02-6600-vector-remap-language-web.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+* The ability to share URLs for specific VRL events and/or programs. This would likely require using
```
Comment (sghall) [inline rfcs/2021-03-02-6600-vector-remap-language-web.md:30]: I think the ability to share is an important one to making this successful but it ups the complexity of the project.  If we want to support that I think we have to support versions so an old link will continue to work.  I think the TS playground is an example where the inputs are encoded in the URL so we don't need to store this in a backend somewhere.  https://www.typescriptlang.org/play

File: rfcs/2021-03-02-6600-vector-remap-language-web.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+1. An in-browser [web app][web_app]/playground for experimenting with and learning VRL
+2. A [web server][web_server] that can process VRL programs and events
+3. A [published Wasm module], largely for internal use
+
+I propose pursuing #1 and #3 in the near term, while leaving #2 as an option for the future.
+
+### Interactive web app
+
+An interactive VRL web app would be a browser-based app that provides an aesthetically pleasing and
```
Comment (sghall) [inline rfcs/2021-03-02-6600-vector-remap-language-web.md:22]: Just want to mention that I think the design of this is related to the upcoming work on the docs site.  We are looking to choose an open source library to replace the theming and components that was coming from TUI.  Whatever we land on there, I think it's going to make sense to use it here as well.

File: rfcs/2021-03-02-6600-vector-remap-language-web.md (hunk 3 above)
Comment (lucperkins) [reply rfcs/2021-03-02-6600-vector-remap-language-web.md:26]: My gut feeling is that the app should only have a concept of "latest" until we can establish that people are deriving real value from it

File: rfcs/2021-03-02-6600-vector-remap-language-web.md (hunk 6 above)
Comment (lucperkins) [reply rfcs/2021-03-02-6600-vector-remap-language-web.md:22]: That makes sense. Let's sync up about this soon, as we'll need to get people from the DD site involved as well, as they have their own imperatives that we'll need to heed.

File: rfcs/2021-03-02-6600-vector-remap-language-web.md (hunk 3 above)
Comment (sghall) [reply rfcs/2021-03-02-6600-vector-remap-language-web.md:26]: > If you build it, they will come.

Yep, I hear ya.  I'm concerned we may not be giving this a fair go by being too conservative on features.   Also thinking about ways we can measure engagement.  That's something we could talk about next week.

Comment (lucperkins) [reply rfcs/2021-03-02-6600-vector-remap-language-web.md:26]: Regarding "being too conservative on features:" now *that* is the kind of can-do gumption we like around here :+1:

**Comment by lucperkins** [comment]
See also #6604

**Pull request closed without merging**